SU2_CFD/src/solvers/CNEMOEulerSolver.cpp

```diff
@@ -2906,13 +2843,130 @@ void CNEMOEulerSolver::BC_Supersonic_Outlet(CGeometry *geometry, CSolver **solut
+      geometry->vertex[val_marker][iVertex]->GetNormal(Normal);
+
+      /*--- Calculate parameters from the geometry ---*/
+      Area = GeometryToolbox::Norm(nDim, Normal);
+
+      for (iDim = 0; iDim < nDim; iDim++){
+        NormalArea[iDim] = -Normal[iDim];
+        UnitNormal[iDim] = -Normal[iDim]/Area;
+      }
+
+      /*--- Retrieve the pressure on the vertex ---*/
+      P   = nodes->GetPressure(iPoint);
+
+      /*--- Apply the flow-tangency b.c. to the convective flux ---*/
+      for (iSpecies = 0; iSpecies < nSpecies; iSpecies++)
+        Residual[iSpecies] = 0.0;
+      for (iDim = 0; iDim < nDim; iDim++){
+        Residual[nSpecies+iDim] = P * UnitNormal[iDim] * Area;
+      }
+      Residual[nSpecies+nDim]   = 0.0;
+      Residual[nSpecies+nDim+1] = 0.0;
+
+      /*--- Add value to the residual ---*/
+      LinSysRes.AddBlock(iPoint, Residual);
+
+      /*--- If using implicit time-stepping, calculate b.c. contribution to Jacobian ---*/
+      if (implicit) {
+
+        /*--- Initialize Jacobian ---*/
+        for (iVar = 0; iVar < nVar; iVar++)
+          for (jVar = 0; jVar < nVar; jVar++)
+            Jacobian_i[iVar][jVar] = 0.0;
+
+        /*--- Calculate state i ---*/
+        rho     = nodes->GetDensity(iPoint);
+        rhoE    = nodes->GetSolution(iPoint,nSpecies+nDim);
+        rhoEve  = nodes->GetSolution(iPoint,nSpecies+nDim+1);
+        dPdU    = nodes->GetdPdU(iPoint);
+        for (iDim = 0; iDim < nDim; iDim++)
+          u[iDim] = nodes->GetVelocity(iPoint,iDim);
+
+        conc = 0.0;
+        for (iSpecies = 0; iSpecies < nSpecies; iSpecies++) {
+          cs    = nodes->GetMassFraction(iPoint,iSpecies);
+          conc += cs * rho/Ms[iSpecies];
+
+          /////// NEW //////
+          for (iDim = 0; iDim < nDim; iDim++) {
+            Jacobian_i[nSpecies+iDim][iSpecies] = dPdU[iSpecies] * UnitNormal[iDim];
+            Jacobian_i[iSpecies][nSpecies+iDim] = cs * UnitNormal[iDim];
+          }
+        }
+
+        for (iDim = 0; iDim < nDim; iDim++) {
+          for (jDim = 0; jDim < nDim; jDim++) {
+            Jacobian_i[nSpecies+iDim][nSpecies+jDim] = u[iDim]*UnitNormal[jDim]
+                + dPdU[nSpecies+jDim]*UnitNormal[iDim];
+          }
+          Jacobian_i[nSpecies+iDim][nSpecies+nDim]   = dPdU[nSpecies+nDim]  *UnitNormal[iDim];
+          Jacobian_i[nSpecies+iDim][nSpecies+nDim+1] = dPdU[nSpecies+nDim+1]*UnitNormal[iDim];
+
+          Jacobian_i[nSpecies+nDim][nSpecies+iDim]   = (rhoE+P)/rho * UnitNormal[iDim];
+          Jacobian_i[nSpecies+nDim+1][nSpecies+iDim] = rhoEve/rho   * UnitNormal[iDim];
+        }
+
+        /*--- Integrate over the dual-grid area ---*/
+        for (iVar = 0; iVar < nVar; iVar++)
+          for (jVar = 0; jVar < nVar; jVar++)
+            Jacobian_i[iVar][jVar] = Jacobian_i[iVar][jVar] * Area;
+
+        /*--- Apply the contribution to the system ---*/
+        Jacobian.AddBlock(iPoint,iPoint,Jacobian_i);
+
```

Comment:
In cases of manipulating the diagonal of the matrix you should always use AddBlock2Diag (or something like that) see CSysMatrix.hpp for the right name.

Reply:
I removed this Euler_Wall_BC and replaced it with the one in **CFVMFlowSolverBase.inl**